src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
+        last = newNode;
+        if (oldLast == null) {
+            first = newNode;
+        } else {
+            oldLast.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> findNode(int index) {
+        if (index < (size >> 1)) {
+            Node<T> temp = first;
+            for (int i = 0; i < index; i++) {
+                temp = temp.next;
+            }
+            return temp;
+        } else {
+            Node<T> temp = last;
+            for (int i = size - 1; i > index; i--) {
+                temp = temp.prev;
+            }
+            return temp;
+        }
+    }
+
+    private void linkBefore(T value, Node<T> node) {
+        Node<T> newNode = new Node<>(node.prev, value, node);
+        if (node.prev == null) {
+            first = newNode;
+        } else {
+            node.prev.next = newNode;
+        }
+        node.prev = newNode;
+        size++;
+    }
+
+    private void checkElementIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("There is no element with this index");
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("This index is invalid");
+        }
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+        final T element = node.item;
+        if (prev == null) {
+            first = node.next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        node.item = null;
+        size--;
+        return element;
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, findNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] arrayElements = list.toArray();
+        Node<T> prev = last;
+        for (Object object : arrayElements) {
+            T element = (T) object;
+            Node<T> newNode = new Node<>(prev, element, null);
+            if (prev == null) {
+                first = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            prev = newNode;
+            size++;
+        }
+        last = prev;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return findNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> temp = findNode(index);
```

Comment:
This is code repeating. You always check the index before getting node by index. What you should do is move the check logic INSIDE the findNode method.